CONTRIBUTING.md

```diff
@@ -1,73 +1,76 @@
+  * Hash-based data structures are usually faster than tree-based data structures. Unless you have a reason to use the latter, prefer unordered_(map|set) over map and set.
+* Copies
+
+    * C++ makes it too easy to inadvertently copy objects.
+
+    * For larger elements (e.g., vectors), pass a (const) reference instead.
+
+    * When your implemented class does not need to be copied (e.g., a Table should never exist more than once), inherit from Noncopyable to avoid these potentially expensive copies.
+
+* Miscellaneous
+  * Prefer `if (object) {` over `if (object != nullptr) {` or `if (object.has_value()) {`.
+  * Don't write `this->` if you don't have to.
+  * Be explicit with types: Use [u]int(8|16|32|64)_t instead of `int, long, uint` etc.
+  * Use auto-to-stick (https://www.fluentcpp.com/2018/09/28/auto-stick-changing-style/): `auto x = 17;` or `auto y = std::vector<size_t>{};`.
+  * Namespaces: Do not create nested namespaces, do not import namespaces.
+
+# Formatting and Naming
+* Much of this is enforced by clang-tidy. However, clang-tidy does not yet cover hpp files (see #1901). Also, while clang-tidy is a great help, do not rely on it.
+* Call ./scripts/format.sh before pushing.
+* Choose clear and concise names, and avoid, e.g., `i`, `j`, `ch_ptr`.
+* Formatting details: 2 spaces, 120 columns, comments above code.
+* Use empty lines to structure your code.
+* Naming conventions:
+
+    * Files: lowercase separated by underscores, e.g., abstract_operator.cpp, usually corresponding to a class, e.g., AbstractOperator.
+
+    * Types (classes, structs, enums, typedefs, using): CamelCase starting with uppercase letter, e.g., `TableScan`.
+
+    * Variables: lowercase separated by underscores, e.g., `chunk_size`.
+
+    * Functions: lowercase separated by underscores, e.g., `append_mutable_chunk()`.
+
+    * Private / protected members / methods: like variables / functions with leading underscore, e.g., `_on_execute()`.
+
+    * Classes that are used only to have a non-templated base class are named `BaseXY` (e.g., `BaseValueSegment`, while classes that have multiple differing implementations are named `AbstractXY` (e.g., `AbstractOperator`).
+
+    * In cases where a constructor parameter would have the same name as the member it initializes, prefix it with `init`: `C(int init_foo) : foo(init_foo) {}`.
+
+    * If an identifier contains a verb or an adjective in addition to a noun, the schema [verb|adjective]\[noun] is preferred, e.g., use `left_input` rather than ~~`input_left`~~ and `set_left_input()` rather than ~~`set_input_left()`~~.
+
+* Maintain correct orthography and grammar. Comments should start with a capital letter, sentences should be finished with a full stop.
+  * Class names within comments are written in CamelCase - e.g., "As the input may be a ReferenceSegment, a valid RowID may *point to* a row that is NULL."
```

Comment:
Maybe add that we use `//` for single line comments, but prefer

```cpp
/**
 * Topic XY:
 *  ...
 */
```

for comment blocks.

Reply:
This is, at least, applied inconsistently. I don't really like the `/* */` style, but would be fine with using it. Before we make this a rule however, we should make sure that the codebase follows it to a large degree.

Honestly, I don't see a big advantage in prescribing something here. For me, code highlighting does the job.

Any other opinions on this?

Reply:
I don't care. If we continue to use the `*` in every line of the comment, I don't see a reason to use multi line comments. It saves one char and is still annoying to write. :)

I am fine with using `//` everywhere and would prefer it (a little bit). I would definitely like to see a strong recommendation in the guidelines for one of both.

Reply:
> I would definitely like to see a strong recommendation in the guidelines for one of both.

> In my opinion, the guidelines should document what we have done so far to keep the code base clean. If we want to introduce something new, it should first be realized in the code base and then added to the guidelines, not the other way round. Otherwise, we risk that people see the guidelines as wishful thinking.

Discussion over here: https://github.com/hyrise/hyrise/pull/2341#discussion_r606756175